### PR TITLE
Add hasFields, Keys, Pluck, Values and Without manipulations #52

### PR DESCRIPTION
**Pull request by Th3Mouk**
## Purpose
#52 

#### Open Questions and Pre-Merge TODOs
- [x] Add tests suite

## Solution

Theses operators aren't tested yet

### CODE REVIEW

**Comment by Th3Mouk** [comment]
@tbolier I'm totally stuck on operators implementation, i don't get the global logic of the implementation. Especially on the `hasFields` one.
The try is about:
```
r.db('test')
  .table('tabletest')
  .filter(r.row.hasFields('number'))
```
And the implementation is in [HasFieldsTest](https://github.com/tbolier/php-rethink-ql/pull/69/files#diff-b23a02d03eb9897f356b3e7956b96acc)

The problems are:
- row() fonction must be without parameters (i think it's cool now)
- filter() operator is not well documented

I don't understand if the `HasFields` operators must return a Row, a QueryInterface, himself, or something else.
If I must create a ManipulationTrait or use the Row classes to add the method. And why here in FilterByRow my manipulation query (HasFields object) is null.

I'm afraid theses problem impact the date operators implementation too. I need to start the dev of a new API and without them I probably will switch to another driver.

File: src/Query/Manipulation/Keys.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+class Keys extends AbstractQuery
+{
+    use AggregationTrait;
+    use OperationTrait;
+    use TransformationTrait;
+
+    /**
+     * @var array
+     */
+    private $keys;
```
Comment (tbolier) [inline src/Query/Manipulation/Keys.php]: Can you remove this unused property?

**Comment by Th3Mouk** [comment]
Yeah please I need some explanations to finish those operators

File: src/Query/Manipulation/HasFields.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
+<?php
+declare(strict_types = 1);
+
+namespace TBolier\RethinkQL\Query\Manipulation;
```
Comment (tbolier) [inline src/Query/Manipulation/HasFields.php]: Please move this to `Query\Logic`

File: src/Query/Manipulation/HasFields.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+use TBolier\RethinkQL\Query\AbstractQuery;
+use TBolier\RethinkQL\Query\Aggregation\AggregationTrait;
+use TBolier\RethinkQL\Query\Operation\OperationTrait;
+use TBolier\RethinkQL\Query\QueryInterface;
+use TBolier\RethinkQL\Query\Transformation\TransformationTrait;
+use TBolier\RethinkQL\RethinkInterface;
+use TBolier\RethinkQL\Types\Term\TermType;
+
+class HasFields extends AbstractQuery
```
Comment (tbolier) [inline src/Query/Manipulation/HasFields.php]: Can you rename the class to `HasFieldsLogic` to keep the naming similar to the existing `GetFieldLogic`?

File: src/Query/Builder.php [hunk 4]
```diff
@@ -70,7 +70,7 @@ public function ordening(string $key): Ordening
         return $this->ordering;
     }
 
-    public function row(string $value): Row
+    public function row(?string $value = null): Row
```
Comment (tbolier) [inline src/Query/Builder.php]: Isn't the `?` is obsolete since `null` probably will never be passed?

**Comment by Th3Mouk** [comment]
No problem ! I prefer a few exemples to long explanations

**Comment by tbolier** [comment]
For `Keys` the implementation was fine, only a nesting query fix and returning concrete types for auto completion at client.

**Comment by tbolier** [comment]
@Th3Mouk I am planning to merge this, are you happy with the outcome and feedback, and do you agree I will merge this?

**Comment by tbolier** [comment]
A good tip to debug is to check the XHR request headers from the admin rethinkDB panel, this will show you the raw JSON query send to RethinkDB, I usually compare this to the PHP raw query that I will copy paste from xDebug.

**Pull request merged**